### PR TITLE
Add rsmi plugin for AMD GPU

### PR DESCRIPTION
**Pull request by miketxli**
Signed-off-by: Mike Li <Tianxinmike.Li@amd.com>

### CODE REVIEW

**Comment by miketxli** [comment]
Install rocm-smi library pkg following
https://rocm-documentation.readthedocs.io/en/latest/Installation_Guide/Installation-Guide.html

Then
cd hwloc
./autogen.sh
./configure --enable-rsmi --enable-plugins='rsmi' CPPFLAGS=-I/opt/rocm/rocm_smi/include/rocm_smi LDFLAGS=-L/opt/rocm/rocm_smi/lib --enable-shared

File: config/hwloc.m4 [hunk 1]
```diff
@@ -1019,6 +1019,39 @@ return clGetDeviceIDs(0, 0, 0, NULL, NULL);
...
+    fi
+    if test "x$hwloc_rsmi_happy" = "xyes"; then
+      tmp_save_CFLAGS="$CFLAGS"
+      CFLAGS="$CFLAGS $HWLOC_RSMI_CFLAGS"
+      tmp_save_LIBS="$LIBS"
+      LIBS="$LIBS $HWLOC_RSMI_LIBS"
+      # AC_CHECK_DECLS([nvmlDeviceGetMaxPcieLinkGeneration],,[:],[[#include <nvml.h>]])
+      # CFLAGS="$tmp_save_CFLAGS"
+      # LIBS="$tmp_save_LIBS"
+    fi
```
Comment (bgoglin) [inline config/hwloc.m4:1038]: If you don't have any specific function to check (something that was added after the first RSMI release and that you want to put under #ifdef in hwloc), remove this entire "if/fi" block.

File: hwloc/topology-rsmi.c [hunk 2]
```diff
@@ -0,0 +1,268 @@
...
+      uint64_t function : 3;
+      uint64_t device : 5;
+      uint64_t bus : 8;
+      uint64_t reserved : 16;
+      uint64_t domain : 32;
+#endif
+    };
+    uint64_t bdfid;
+  };
+} rsmiPciInfo_t;
```
Comment (bgoglin) [inline hwloc/topology-rsmi.c:41]: Looks like we need to add some configury to define HWLOC_BIGENDIAN?
Also isn't this structure defined in RSMI headers? I'd be better not to duplicate it in hwloc.

File: hwloc/topology-rsmi.c [hunk 3]
```diff
@@ -0,0 +1,268 @@
...
+ * size			(OUT) Size of the name
+ */
+static void get_device_name(uint32_t dv_ind, char *device_name, unsigned int size)
+{
+  const char *status_string;
+  rsmi_status_t rsmi_rc = rsmi_dev_name_get(dv_ind, device_name, size);
+
+  if (rsmi_rc != RSMI_STATUS_SUCCESS) {
+    rsmi_rc = rsmi_status_string(rsmi_rc, &status_string);
+    fprintf(stderr, "Failed to get name of the GPU: %s\n", status_string);
```
Comment (bgoglin) [inline hwloc/topology-rsmi.c:57]: Make sure these error messages only occur under catastrophic circumstance sicne we don't want the hwloc library to pollute the stderr of programs that don't even know there are using hwloc.
Also add a ref to RSMI and/or rsmi_dev_name() in the message.
Same for all other fprintf(stderr...) below.

File: hwloc/topology-rsmi.c [hunk 4]
```diff
@@ -0,0 +1,268 @@
...
+    get_device_pci_info(i, &pci);
+    parent = hwloc_pci_find_parent_by_busid(topology, pci.domain, pci.bus, pci.device, 0);
+    get_device_pci_linkspeed(i, &(parent->attr->pcidev.linkspeed));
+    if (!parent)
+      parent = hwloc_get_root_obj(topology);
+
+    hwloc_insert_object_by_parent(topology, parent, osdev);
+  }
+
+  rsmi_shut_down();
```
Comment (bgoglin) [inline hwloc/topology-rsmi.c:213]: What happens if the program does its own rsmi_init() either earlier or before? do you have proper reference counting inside RSMI to make sure this rsmi_shut_down() in hwloc shut down RSMI for the entire process?

File: hwloc/topology-rsmi.c [hunk 5]
```diff
@@ -0,0 +1,268 @@
...
+  backend->discover = hwloc_rsmi_discover;
+  return backend;
+}
+
+static struct hwloc_disc_component hwloc_rsmi_disc_component = {
+  "rsmi",
+  HWLOC_DISC_PHASE_IO,
+  HWLOC_DISC_PHASE_GLOBAL,
+  hwloc_rsmi_component_instantiate,
+  5, /* after pci, and after cuda since likely less useful */
```
Comment (bgoglin) [inline hwloc/topology-rsmi.c:239]: Do you really want this ref to CUDA being more useful ? :)
If you want to compare to other IO backends, opencl might be the relevant one since it may also add osdevices for the same GPUs. Do you care whether rsmi is before or after OpenCL? If yes, adjust the priority above or below OpenCL.
Otherwise, keep the same priority as CUDA etc.

File: hwloc/topology-rsmi.c [hunk 6]
```diff
@@ -0,0 +1,268 @@
...
+hwloc_rsmi_component_init(unsigned long flags)
+{
+  if (flags)
+    return -1;
+  if (hwloc_plugin_check_namespace("rsmi", "hwloc_backend_alloc") < 0)
+    return -1;
+  if (!dlopen("librocm_smi64.so", RTLD_NOW | RTLD_GLOBAL)) {
+    fprintf(stderr, "RSMI configured, but wasn't found.\n");
+    return -1;
+  }
```
Comment (bgoglin) [inline hwloc/topology-rsmi.c:254]: Why do you need this dlopen? The rsmi hwloc plugin won't be loaded unless this librocm_smi64.so is available since it will depend on it.

File: config/hwloc.m4 [hunk 7]
```diff
@@ -1019,6 +1019,39 @@ return clGetDeviceIDs(0, 0, 0, NULL, NULL);
     fi
     # don't add LIBS/CFLAGS/REQUIRES yet, depends on plugins
 
+    # RSMI support
+    hwloc_rsmi_happy=no
+    if test "x$enable_io" != xno && test "x$enable_rsmi" != "xno"; then
+      hwloc_rsmi_happy=yes
+      AC_CHECK_HEADERS([rocm_smi.h], [
+        AC_CHECK_LIB([rocm_smi64], [rsmi_init], [HWLOC_RSMI_LIBS="-lrocm_smi64"], [hwloc_rsmi_happy=no])
+        ], [hwloc_rsmi_happy=no])
```
Comment (bgoglin) [inline config/hwloc.m4:1028]: Is there a librocm_smi32.so for 32bits systems?
Add a comment so that we know why we're only looking for the 64bits lib without checking if we're on a 64bit machine.

File: include/hwloc/rsmi.h [hunk 8]
```diff
@@ -0,0 +1,214 @@
...
+      uint64_t function : 3;
+      uint64_t device : 5;
+      uint64_t bus : 8;
+      uint64_t reserved : 16;
+      uint64_t domain : 32;
+#endif
+    };
+    uint64_t bdfid;
+  };
+} rsmiPciInfo_t;
```
Comment (bgoglin) [inline include/hwloc/rsmi.h:64]: If we keep this structure in hwloc as discussed above, remove it from topology-rsmi.c and #include hwloc/rsmi.h in topology-rsmi.c
Also we'll need HWLOC_BIGENDIAN to be exported to the *public* autogen/config.h since application will need the #define when they'll use hwloc/rsmi.h.
(I can take care of HWLOC_BIGENDIAN).

File: include/hwloc/rsmi.h [hunk 9]
```diff
@@ -0,0 +1,214 @@
...
+
+    info = hwloc_obj_get_info_by_name(osdev, "AMDUUID");
+    if (info && !strcmp(info, uuid))
+      return osdev;
+  }
+
+  return NULL;
+}
+
+/** @} */
```
Comment (bgoglin) [inline include/hwloc/rsmi.h:206]: There was some idea of moving some of these helpers to somewhere else. If it's not a priority for you, I'd rather keep them in a separate commit so that we can merge the core backend and discuss helpers laters.

File: hwloc/topology-rsmi.c [hunk 10]
```diff
@@ -0,0 +1,268 @@
...
+    char buffer[64];
+
+    osdev = hwloc_alloc_setup_object(topology, HWLOC_OBJ_OS_DEVICE, HWLOC_UNKNOWN_INDEX);
+    snprintf(buffer, sizeof(buffer), "rsmi%u", i);
+    osdev->name = strdup(buffer);
+    osdev->depth = HWLOC_TYPE_DEPTH_UNKNOWN;
+    osdev->attr->osdev.type = HWLOC_OBJ_OSDEV_GPU;
+
+    hwloc_obj_add_info(osdev, "Backend", "RSMI");
+    hwloc_obj_add_info(osdev, "GPUVendor", "Advanced Micro Devices");
```
Comment (bgoglin) [inline hwloc/topology-rsmi.c:187]: Would "AMD" be acceptable here? This kind of string is usually printed in verbose message, so having something short and easier to recognize wouldn't hurt.

File: include/hwloc/rsmi.h [hunk 11]
```diff
@@ -0,0 +1,214 @@
...
+
+    if (pcidev
+      && pcidev->type == HWLOC_OBJ_PCI_DEVICE
+      && pcidev->attr->pcidev.domain == pci.domain
+      && pcidev->attr->pcidev.bus == pci.bus
+      && pcidev->attr->pcidev.dev == pci.device
+      && pcidev->attr->pcidev.func == pci.function)
+      return osdev;
+
+    info = hwloc_obj_get_info_by_name(osdev, "AMDUUID");
```
Comment (bgoglin) [inline include/hwloc/rsmi.h:198]: I don't know if NVIDIAUUID was a good idea. Maybe just use UUID here?

File: tests/hwloc/ports/include/rsmi/rsmi.h [hunk 12]
```diff
@@ -0,0 +1,68 @@
...
+ */
+
+#ifndef HWLOC_PORT_RSMI_RSMI_H
+#define HWLOC_PORT_RSMI_RSMI_H
+
+/* we need to replace any RSMI-related #define that configure may have put in private/autogen/config.h */
+#ifndef HWLOC_CONFIGURE_H
+#error rsmi.h must be included after private/autogen/config.h
+#endif
+
```
Comment (bgoglin) [inline tests/hwloc/ports/include/rsmi/rsmi.h:15]: This file should be named exactly as the official headers used in topology-rsmi.c so I guess you should replace rsmi/rsmi.h with rocm_smi.h. That's likely the reason why our CI failed.

**Comment by miketxli** [comment]
Thanks, Brice

Let me go through the comments and come back to you.

Thanks

Mike

File: config/hwloc.m4 (hunk 1 above)
Comment (miketxli) [reply config/hwloc.m4:1038]: will remove it.

File: hwloc/topology-rsmi.c (hunk 2 above)
Comment (miketxli) [reply hwloc/topology-rsmi.c:41]: This structure isn't defined in RSMI headers. We can remove HWLOC_BIGENDIAN, since it's not used.

File: hwloc/topology-rsmi.c (hunk 3 above)
Comment (miketxli) [reply hwloc/topology-rsmi.c:57]: OK

File: hwloc/topology-rsmi.c (hunk 2 above)
Comment (bgoglin) [reply hwloc/topology-rsmi.c:41]: Oh I see, only the bdfid part of union seems to be passed to rsmi_dev_pci_id_get(), the struct is here for decoding only. Does RSMI work on some big endian architectures?

File: hwloc/topology-rsmi.c (hunk 4 above)
Comment (miketxli) [reply hwloc/topology-rsmi.c:213]: The first time init() is called, the internal data structures are initialized, and the first time shutdown is called they are freed.

Comment (miketxli) [reply hwloc/topology-rsmi.c:213]: There is no reference counting inside RSMI right now, can be added later.

File: hwloc/topology-rsmi.c (hunk 2 above)
Comment (miketxli) [reply hwloc/topology-rsmi.c:41]: Yes, that's for decoding only. I have ashed some regarding big endian architectures.

File: hwloc/topology-rsmi.c (hunk 4 above)
Comment (bgoglin) [reply hwloc/topology-rsmi.c:213]: It *must* be added as soon as possible. Missing support for multiple init/shutdown is very annoying for developers. Right now, it means hwloc just cannot call shutdown at all: if an application (e.g using hwloc/rsmi.h) calls rsmi_init() before hwloc_topology_load(), hwloc would shutdown rsmi at the end of that hwloc_topology_load().

Not calling shutdown means valgrind will complain. Hence we'll end up adding yet another environment variable sayign "OK hwloc, you can shutdown rsmi, I know what I am doing" so that our regression tests don't fail.

Comment (miketxli) [reply hwloc/topology-rsmi.c:213]: Let me talk to the developer.

File: tests/hwloc/ports/include/rsmi/rsmi.h (hunk 12 above)
Comment (bgoglin) [reply tests/hwloc/ports/include/rsmi/rsmi.h:15]: Fix what I said above: the file should be in include/rsmi/rocm_smi.h (headers used to simulate rsmi build without rsmi are in include/rsmi/)
You may test this part with make check -C tests/hwloc/ports

File: hwloc/topology-rsmi.c (hunk 2 above)
Comment (bgoglin) [reply hwloc/topology-rsmi.c:41]: Given that the rsmi doc explicitly says:
```
 *      BDFID = ((DOMAIN & 0xffffffff) << 32) | ((BUS & 0xff) << 8) |
 *                                   ((DEVICE & 0x1f) <<3 ) | (FUNCTION & 0x7)
```
I would just remove the structure entirely and decode the uint64_t manually, it won't depend on big/little endian.

File: hwloc/topology-rsmi.c (hunk 4 above)
Comment (bgoglin) [reply hwloc/topology-rsmi.c:213]: I put all details in a github issue against rsmi https://github.com/RadeonOpenCompute/rocm_smi_lib/issues/62

**Comment by bgoglin** [comment]
Should we also call rsmi_dev_memory_total_get() to annotate the hwloc object with information like we do for opencl and cuda?
OpenCLGlobalMemorySize=3087360
OpenCLComputeUnits=68
CUDAGlobalMemorySize=16548352
CUDAL2CacheSize=6144
CUDAMultiProcessors=80
CUDACoresPerMP=64
CUDASharedMemorySizePerMP=48

File: hwloc/topology-rsmi.c (hunk 2 above)
Comment (miketxli) [reply hwloc/topology-rsmi.c:41]: Ok

File: hwloc/topology-rsmi.c (hunk 5 above)
Comment (miketxli) [reply hwloc/topology-rsmi.c:239]: Will put it before OpenCL

File: include/hwloc/rsmi.h (hunk 8 above)
Comment (miketxli) [reply include/hwloc/rsmi.h:64]: Will use decode

File: include/hwloc/rsmi.h (hunk 9 above)
Comment (miketxli) [reply include/hwloc/rsmi.h:206]: agree

File: hwloc/topology-rsmi.c (hunk 10 above)
Comment (miketxli) [reply hwloc/topology-rsmi.c:187]: OK

File: include/hwloc/rsmi.h (hunk 11 above)
Comment (miketxli) [reply include/hwloc/rsmi.h:198]: OK

File: tests/hwloc/ports/include/rsmi/rsmi.h (hunk 12 above)
Comment (miketxli) [reply tests/hwloc/ports/include/rsmi/rsmi.h:15]: You mean it should be in tests/hwloc/ports/include/rsmi directory, but change file name to rocm_smi.h same as in topology-rsmi.c?

Comment (bgoglin) [reply tests/hwloc/ports/include/rsmi/rsmi.h:15]: Yes!

File: hwloc/topology-rsmi.c (hunk 4 above)
Comment (miketxli) [reply hwloc/topology-rsmi.c:213]: I will forward it to the developer for rocm_smi library. Like I said at beginning, this is just phase 1 for hwloc plugin. It will be added in phase 2.

**Comment by miketxli** [comment]
We can add those in phase 2.

**Comment by bgoglin** [comment]
OK
Feel free to add some words in doc/hwloc.dox (when we list dependencies, info attributes, OS device names, and interoperability helpers). Look for NVIDIA and add RSMI things nearby :)
I still haven't figured out whether it's better to name your info attribute AMDFoo or just a generic Foo.

File: config/hwloc.m4 (hunk 7 above)
Comment (miketxli) [reply config/hwloc.m4:1028]: There is no  librocm_smi32.so, they call library  librocm_smi64.so

File: hwloc/topology-rsmi.c (hunk 6 above)
Comment (miketxli) [reply hwloc/topology-rsmi.c:254]: Let see hwloc is build in another machine. But rocm_smi isn't installed in the machine hwloc is running, we stop here instead crash when with init.

Comment (bgoglin) [reply hwloc/topology-rsmi.c:254]: If topology-rsmi.c is built statically inside libhwloc (not as plugin), libhwloc.so will depend on librocm_smi64.so explicitly, hence libhwloc.so will fail to load, and the program will fail to start because of the missing librocm64_smi.so
If built as a plugin, libhwloc.so will work fine, the program will start. When hwloc will load plugins, libhwloc_rsmi.so will fail to load because it needs librocm_smi64.so. That's actually the whole point of plugins: build everything but don't require all dependencies to be available at runtime, only those available will have their hwloc plugins loaded.
It'll work fine without dlopen, we already do that for multiple other dependencies.

File: config/hwloc.m4 (hunk 7 above)
Comment (bgoglin) [reply config/hwloc.m4:1028]: Indeed the CMakeLists.txt only enables 64bits by default, and manual changes to switch to 32bits don't touch the library name. Please put a comment about this :)

File: hwloc/topology-rsmi.c (hunk 6 above)
Comment (miketxli) [reply hwloc/topology-rsmi.c:254]: Ok

File: config/hwloc.m4 (hunk 7 above)
Comment (miketxli) [reply config/hwloc.m4:1028]: will add a conmments.

**Comment by miketxli** [comment]
I will go ahead and make update based on what we discussed. Are you OK with that?

**Comment by bgoglin** [comment]
Yes sure, go ahead. You may even force push to replace your previous commit with a better one.

**Comment by miketxli** [comment]
I'm multitasking right now. It will take some time for me to come back.
Thanks

Mike

**Comment by miketxli** [comment]
Sorry about delay. I'm working on this task now. The good news is that they have added a ref count for rocm-smi library. 
I have an issue after merged with latest code. 
It used to show plugin (rsmi), now shows dlopen(rsmi) after configuration. When I run lstopo, I got segmentation fault. It has something to do with recent change for AIX. I'm debug it. Maybe you can point me to right direction.

Thanks

Mike

**Comment by miketxli** [comment]
I fixed the issue by disable dlopen.

Thanks

Mike

**Comment by bgoglin** [comment]
Can you clarify what you meant by "recent change for AIX"?
I review the code next week, thanks.

**Comment by bgoglin** [comment]
CI failed, looks like you need to add rsmi_status_string to tests/hwloc/ports/include/rsmi/rocm_smi.h
```
In file included from topology-rsmi.c:17:
/builds/workspace/basic_PR-398/hwloc-PR-398-20200522.2236.git60689bb/tests/hwloc/ports/include/rsmi/rocm_smi.h:37: error: ‘RSMI_MAX_NUM_FREQUENCIES’ undeclared here (not in a function)
topology-rsmi.c: In function ‘get_device_name’:
topology-rsmi.c:32: warning: implicit declaration of function ‘rsmi_status_string’
```

**Comment by miketxli** [comment]
It's related to the following check in 
commit 9c02aa1a76d2422cc13d38ddaddfcc23087fe8aa
Author: Brice Goglin <Brice.Goglin@inria.fr>
Date:   Mon Feb 24 17:11:46 2020 +0100

    configure: use dlopen instead of ltdl by default
    
I have to disable dlopen to use ltdl.

**Comment by miketxli** [comment]
I fix the compile issue.

**Comment by bgoglin** [comment]
Which Linux distribution do you use? I'll see if I can test the dlopen code on a similar one here.

**Comment by miketxli** [comment]
Ubuntu 18.04, I have to do the following:
./configure --disable-plugin-dlopen --enable-rsmi --enable-plugins='rsmi' CPPFLAGS="-I/opt/rocm/rocm_smi/include/rocm_smi -I/opt/rocm/rocm_smi/include" LDFLAGS=-L/opt/rocm/rocm_smi/lib --enable-shared
to disable dlopen

**Comment by bgoglin** [comment]
We're trying to reproduce the dlopen issue here, and installing RSMI to test the code. In the meantime, here are a couple minor changes:
* Remove #include dlfcn.h
* Change all small get_device_foo() helper function to return 0/-1 on success error, or set the buffer to sane default values on error (e.g. set the linkspeed to 0).
* Change back UUID to AMDUUID since you still have AMDSerial, and we cannot easily change NVIDIA info key names anymore. Sorry for changing my mind here.

**Comment by miketxli** [comment]
Have updated base on the comments.
thanks
Mike

**Comment by bgoglin** [comment]
Thanks. Is the kernel driver in the upstream Linux kernel? Which minimal kernel version do we need? We couldn't reproduce the dlopen crash so far, but rsmi_init() fails because of the missing kernel driver :)

File: config/hwloc_internal.m4 [hunk 13]
```diff
@@ -93,6 +93,11 @@ AC_DEFUN([HWLOC_DEFINE_ARGS],[
                   AS_HELP_STRING([--disable-nvml],
                                  [Disable the NVML device discovery build (instead of only disabling NVML at runtime by default)]))
 
+    # RSMI?
+    AC_ARG_ENABLE([rsmi],
+                  AS_HELP_STRING([--disable-rsmi],
+                                 [Disable the RSMI device discovery]))
+
```
Comment (bgoglin) [inline config/hwloc_internal.m4:100]: Can you write something more clear such "ROCm SMI" in this documentation line?

File: tests/hwloc/ports/include/rsmi/rocm_smi.h [hunk 14]
```diff
@@ -0,0 +1,69 @@
...
+ * See COPYING in top-level directory.
+ */
+
+#ifndef HWLOC_PORT_RSMI_RSMI_H
+#define HWLOC_PORT_RSMI_RSMI_H
+
+/* we need to replace any RSMI-related #define that configure may have put in private/autogen/config.h */
+#ifndef HWLOC_CONFIGURE_H
+#error rsmi.h must be included after private/autogen/config.h
+#endif
```
Comment (bgoglin) [inline tests/hwloc/ports/include/rsmi/rocm_smi.h:14]: This #error should say rocm_smi.h instead of rsmi.h

File: config/hwloc_internal.m4 (hunk 13 above)
Comment (miketxli) [reply config/hwloc_internal.m4:100]: done

File: tests/hwloc/ports/include/rsmi/rocm_smi.h (hunk 14 above)
Comment (miketxli) [reply tests/hwloc/ports/include/rsmi/rocm_smi.h:14]: Ok, good finding

**Comment by miketxli** [comment]
We should be able to use any version of driver, but lets try with latest up steamed.

**Comment by bgoglin** [comment]
I am preparing some documentation updates. They raised 2 questions:
* did you consider using OSDEV subtype COPROC instead of GPU? The difference is small but I assume ROCm is not only about graphics but also about computing. CUDA uses COPROC. NVML uses GPU because it's really for management, and they create separate OS devices. I'd like to cleanup those subtypes in the future but it's not feasible for now
* I found the Debian packages name 'librocm-smi-dev" (Debian will ship official ROCm packages soon) but I am not sure about RPM package names. I found rocm_smi.h in "rocm-smi-lib64" in the AMD repo but this name looks a bit awkward to me. Do you know if that's what RHEL might ship in the future?

**Comment by miketxli** [comment]
For the first question, I would use GPU. RSMI is used to manager GPU, similar to NVML. The compute API is provided by HIP, similar to CUDA.
For the second question, I will ask the developer and come back to you.

File: tests/hwloc/ports/include/rsmi/rocm_smi.h (hunk 14 above)
Comment (bgoglin) [reply tests/hwloc/ports/include/rsmi/rocm_smi.h:14]: typo in rsmi_smi.h :)

**Comment by bgoglin** [comment]
Here's the doc update I was talking about. Feel free to include them in your commit, or I 'll push them after merging your PR.
https://github.com/bgoglin/hwloc/commit/6c501ad8f7d53a856ba62e1c15f242026cec9ded

**Comment by miketxli** [comment]
Updated based on your last comments.
For 1.st question, I have asked some high level architects and is waiting for comments.
For 2nd question,  it should be the same for both deb and rpm.

Thanks

Mike

**Comment by bgoglin** [comment]
I am finally testing this on a real machine with a radeon fiji nano. The plugin works fine with dlopen here on Centos7 with rocm-smi-lib64.x86_64 package version 2.2.0.3.rocm_rel_3.3_19_a482394-1

get_uuid fails (I added a status string printf but it's NULL).
get_serial returns an empty string.
name, linkspeed and BDF look correct.

Can you change the code to do these?
* not set AMDSerial and AMDUUID when you get an error (instead of setting to empty string)
* fix the if for PCI BDF in the main function, it should check for ==0 instead of -1
* not set the BFD to 0 when you get an error. just set parent to hwloc_get_root_obj(topology); too, in that case.
* by the way, the status string variable may move to inside the ifs
* why do you printf in some cases and not others? people running apps over MPI over hwloc might wonder why they get these error even if they don't care about hwloc. The core hwloc has a way to hide such messages when HWLOC_HIDE_ERRORS=1 in the env (but it's not available to plugins, so we'd to duplicate getenv here).

I can take care of these changes now that I can test things. But I'd need your help about why rocm_smi return empty serial and fails to return the uuid.

**Comment by miketxli** [comment]
get_uuid fails (I added a status string printf but it's NULL).
get_serial returns an empty string.
uuid and serial are supported only by newer GPU's, Fiji doesn't support it.

I made fix for other things. You can change it if it's minor.

Thanks

Mike

File: hwloc/topology-rsmi.c [hunk 15]
```diff
@@ -0,0 +1,243 @@
...
+    if (get_device_pci_info(i, &bdfid) == 0) {
+      domain = (bdfid>>32) & 0xffffffff;
+      bus = ((bdfid & 0xffff)>>8) & 0xff;
+      device = ((bdfid & 0xff)>>3) & 0x1f;
+      func = bdfid & 0x7;
+      parent = hwloc_pci_find_parent_by_busid(topology, domain, bus, device, func);
+      if (parent) {
+        if (get_device_pci_linkspeed(i, &(parent->attr->pcidev.linkspeed)) == -1)
+          parent->attr->pcidev.linkspeed = 0;
+      } else {
```
Comment (bgoglin) [inline hwloc/topology-rsmi.c:184]: add "&& parent->type == HWLOC_OBJ_PCI_DEVICE" in "if (parent)".

**Comment by bgoglin** [comment]
I think I am ready to merge this beside the small change I just put in the review.
Is it ready to merge on your side too? After this, I'll add envvars HWLOC_HIDE_ERRORS=1 (to optionally silence error messages) and HWLOC_RSMI_NO_SHUTDOWN=1 (to optionally disable rsmi_shutdown() if it breaks things when running a librsmi without refcounting).

File: hwloc/topology-rsmi.c [hunk 16]
```diff
@@ -0,0 +1,243 @@
...
+
+    buffer[0] = '\0';
+    if (get_device_name(i, buffer, sizeof(buffer)) == -1)
+      buffer[0] = '\0';
+    hwloc_obj_add_info(osdev, "GPUModel", buffer);
+
+    buffer[0] = '\0';
+    if (get_device_serial_number(i, buffer, sizeof(buffer)) == 0)
+      hwloc_obj_add_info(osdev, "AMDSerial", buffer);
+
```
Comment (bgoglin) [inline hwloc/topology-rsmi.c:169]: if buffer can be left empty, add "&& buffer[0]" at the end of the if() above

**Comment by miketxli** [comment]
Added type check. If you ready, you can merge. I don't have more to change.

Thanks

Mike

**Comment by bgoglin** [comment]
I just put yet another comment in review. Otherwise testing looks good here.

File: hwloc/topology-rsmi.c (hunk 16 above)
Comment (miketxli) [reply hwloc/topology-rsmi.c:169]: buffer can contain some data, even so call failed.

Comment (bgoglin) [reply hwloc/topology-rsmi.c:169]: What I am getting on Fiji is not a failure but an empty buffer, so I wanted to ignore this empty AMDSerial.

Comment (miketxli) [reply hwloc/topology-rsmi.c:169]: I didn't that. Ok, I will add check.

Thanks

Mike

**Comment by bgoglin** [comment]
@miketxli Now that we merged this, I seem to remember you had a hwloc/rsmi.h and a test case in the first submission? Can you push them back to a branch so that we look at them now?

**Comment by miketxli** [comment]
Ok, let me take a look. Can you tell me how to compile and test that?

**Comment by bgoglin** [comment]
Tarballs of git master are updated here after each commit/push https://ci.inria.fr/hwloc/job/basic/job/master/

**Comment by bgoglin** [comment]
@miketxli Hello, 2 questions.

hwloc/rsmi doesn't build over a manually-built rocm_smi_lib from github master because rocm_smi.h now includes rocm_smi/kfd_ioctl.h. It looks like to include rocm_smi/rocm_smi.h instead of rocm_smi.h (and remove the ending "rocm_smi/" from the include path). Can you confirm this?

I tried on a machine with 2 tahiti GPUs, rocm_smi doesn't seem to report any GPU. Are these cards too old?

**Comment by miketxli** [comment]
For 1.st question, yes, they have added one more header file with another path. I have made it work by doing this: 
./configure --disable-plugin-dlopen --enable-rsmi --enable-plugins='rsmi' CPPFLAGS="-I/opt/rocm/rocm_smi/include/rocm_smi -I/opt/rocm/rocm_smi/include" LDFLAGS=-L/opt/rocm/rocm_smi/lib --enable-shared

For 2nd question, I haven't tested with any GPU older than Fiji. I will get the information for you.

Thanks

Mike

**Comment by miketxli** [comment]
For 1.st question, they will take out the extra include later.
For 2nd question, the tahiti GPUs may not be supported by default through the kernel driver. How many AMD cards do you have? If you need newer cards, AMD can send you some sample cards.

**Comment by bgoglin** [comment]
1) Given that rocm_smi_lib uses rocm_smi/rocm_smi.h everywhere, including in its own example, I changed the code to do the same. So now we configure with rocm_smi/include instead of rocm_smi/include/rocm_smi
2) We only have 2 very old tahiti (bought in 2012) and 2 fiji nano xt (sent by AMD a couple years ago when we started looking at HSA). If AMD can send some newer cards, it would be great. We'll need cards whose cooling is compatible with rack servers.

**Comment by bgoglin** [comment]
Hello @miketxli Don't you have your original submission somewhere where you included hwloc/rsmi.h and tests/hwloc/rsmi.c? I will be able to test those now.

**Comment by miketxli** [comment]
I have created pull request 404, have tested under Ubuntu 18.04
